Upgrade connexion package to version 3.0

mhamra opened the pull request:
|Related issue|
|---|
|#16504|

## Description
This PR contains the required changes to upgrade the connexion package to version 3.0,
This new release has breaking changes that involve a lot of adjustments to the existing code. AioHTTP was replaced by an ASGI server and is implemented on top of the starlette framework.

nico-stefani commented on framework/requirements.txt:
```diff
@@ -1,12 +1,6 @@
 asn1crypto==1.3.0
 async-timeout==4.0.2
-attrs==20.3.0
+attrs=23.1.0
```
```suggestion
attrs==23.1.0
```

nico-stefani commented on framework/wazuh/rbac/auth_context.py:
```diff
@@ -58,7 +58,7 @@ def __init__(self, auth_context: Union[dict, str] = None, role: Union[list, orm.
             # All system's roles
             with orm.RolesManager() as rm:
-                roles_list = map(orm.Roles.to_dict, rm.get_roles())
+                roles_list = list(map(orm.Roles.to_dict, rm.get_roles()))
```
Why this change is needed?

mhamra replied:
This change is the solution found by @fdalmaup [here](https://github.com/wazuh/wazuh/pull/20527/files) to the warning messages produced by SQLAlchemy. It was helpful to include this fix to avoid long warning messages in the console log when I was debugging the API execution.

It can be discarded and merged in the future or maintained.

nico-stefani reviewed (changes requested):
Could you add the output of the AIT execution?

nico-stefani commented on framework/requirements.txt:
```diff
@@ -44,19 +38,19 @@ grpc-google-iam-v1==0.12.3
 Jinja2==3.0.0
 jmespath==0.9.5
-jsonschema==2.6.0
+jsonschema=4.20.0
```
```suggestion
jsonschema==4.20.0
```

mhamra commented on api/test/integration/env/base/manager/manager.Dockerfile:
```diff
@@ -12,6 +12,7 @@ ADD base/manager/supervisord.conf /etc/supervisor/conf.d/
 RUN mkdir wazuh && curl -sL https://github.com/wazuh/wazuh/tarball/${WAZUH_BRANCH} | tar zx --strip-components=1 -C wazuh
 COPY base/manager/preloaded-vars.conf /wazuh/etc/preloaded-vars.conf
 RUN /wazuh/install.sh
+RUN /var/ossec/framework/python/bin/pip3 install -r /wazuh/framework/requirements.txt
```
This change is required to run the AITs until a new embedded version includes the packages in the `requirements.txt` file.

mhamra commented on api/test/integration/env/docker-compose.yml:
They are just for local execution but have no conflict to be kept.

mhamra commented on api/api/authentication.py:
```diff
@@ -275,7 +276,7 @@ def decode_token(token: str) -> dict:
 
         if not data['result']['valid']:
-            raise Unauthorized
+            raise Unauthorized("Invalid token")
```
OK, constant defined.

mhamra commented on api/api/controllers/cdb_list_controller.py:
```diff
@@ -100,11 +99,11 @@ async def get_file(request, pretty: bool = False, wait_for_complete: bool = Fals
-        Depending on the `raw` parameter, it will return a web.Response object or a ConnexionResponse object:
+    ConnexionResponse or ConnexionResponse
+        Depending on the `raw` parameter, it will return a ConnexionResponse object or a ConnexionResponse object:
             raw=True            -> ConnexionResponse (text/plain)
-            raw=False (default) -> web.Response      (application/json)
-        If any exception was raised, it will return a web.Response with details.
+            raw=False (default) -> ConnexionResponse      (application/json)
+        If any exception was raised, it will return a ConnexionResponse with details.
```
Fixed in a new commit.

mhamra commented on api/api/controllers/cluster_controller.py:
```diff
@@ -366,11 +358,11 @@ async def get_configuration_node(request, node_id: str, pretty: bool = False, wa
-        Depending on the `raw` parameter, it will return a web.Response object or a ConnexionResponse object:
+    ConnexionResponse or ConnexionResponse
+        Depending on the `raw` parameter, it will return a ConnexionResponse object or a ConnexionResponse object:
             raw=True            -> ConnexionResponse (application/xml)
-            raw=False (default) -> web.Response (application/json)
-        If any exception was raised, it will return a web.Response with details.
+            raw=False (default) -> ConnexionResponse (application/json)
+        If any exception was raised, it will return a ConnexionResponse with details.
```
OK, fixed in a new commit

mhamra commented on api/api/controllers/decoder_controller.py:
```diff
@@ -236,10 +232,10 @@ async def get_file(request, pretty: bool = False, wait_for_complete: bool = Fals
-        Depending on the `raw` parameter, it will return a web.Response object or a ConnexionResponse object:
+        Depending on the `raw` parameter, it will return a ConnexionResponse object or a ConnexionResponse object:
             raw=True            -> ConnexionResponse (application/xml)
-            raw=False (default) -> web.Response (application/json)
-        If any exception was raised, it will return a web.Response with details.
+            raw=False (default) -> ConnexionResponse (application/json)
+        If any exception was raised, it will return a ConnexionResponse with details.
```
OK, fixed in a new commit.

mhamra commented on api/api/controllers/event_controller.py:
```diff
@@ -4,19 +4,20 @@
 
-from aiohttp import web
+from connexion.lifecycle import ConnexionResponse
 from wazuh.core.cluster.dapi.dapi import DistributedAPI
 from wazuh.event import send_event_to_analysisd
 
-from api.encoder import dumps, prettify
+from connexion import request
+from api.controllers.util import json_response, JSON_CONTENT_TYPE
```
OK, reordered in a new commit.

mhamra commented on api/api/controllers/experimental_controller.py:
```diff
@@ -5,14 +5,15 @@
 
-from aiohttp import web
+from connexion.lifecycle import ConnexionResponse
 
 import wazuh.ciscat as ciscat
 import wazuh.rootcheck as rootcheck
 import wazuh.syscheck as syscheck
 import wazuh.syscollector as syscollector
 from api import configuration
-from api.encoder import dumps, prettify
+from connexion import request
+from api.controllers.util import json_response
 from api.util import remove_nones_to_dict, parse_api_param, raise_if_exc
```
OK, reordered in a new commit.

mhamra commented on api/api/controllers/manager_controller.py:
```diff
@@ -107,11 +104,11 @@ async def get_configuration(request, pretty: bool = False, wait_for_complete: bo
-        Depending on the `raw` parameter, it will return a web.Response object or a ConnexionResponse object:
+    ConnexionResponse or ConnexionResponse
+        Depending on the `raw` parameter, it will return a ConnexionResponse object or a ConnexionResponse object:
             raw=True            -> ConnexionResponse (application/xml)
-            raw=False (default) -> web.Response (application/json)
-        If any exception was raised, it will return a web.Response with details.
+            raw=False (default) -> ConnexionResponse (application/json)
+        If any exception was raised, it will return a ConnexionResponse with details.
```
OK, fixed in a new commit.

mhamra commented on api/api/controllers/overview_controller.py:
```diff
@@ -4,30 +4,30 @@
 
-from aiohttp import web
+from connexion.lifecycle import ConnexionResponse
 
-from api.encoder import dumps, prettify
+from connexion import request
+from api.controllers.util import json_response
```
OK, reordered in a new commit.

mhamra commented on api/api/controllers/rule_controller.py:
```diff
@@ -299,11 +287,11 @@ async def get_file(request, pretty: bool = False, wait_for_complete: bool = Fals
-        Depending on the `raw` parameter, it will return a web.Response object or a ConnexionResponse object:
+    ConnexionResponse or ConnexionResponse
+        Depending on the `raw` parameter, it will return a ConnexionResponse object or a ConnexionResponse object:
             raw=True            -> ConnexionResponse (application/xml)
-            raw=False (default) -> web.Response      (application/json)
-        If any exception was raised, it will return a web.Response with details.
+            raw=False (default) -> ConnexionResponse      (application/json)
+        If any exception was raised, it will return a ConnexionResponse with details.
     """
```
OK, fixed in a new commit.

mhamra commented on api/api/controllers/sca_controller.py:
```diff
@@ -5,26 +5,26 @@
 
-from aiohttp import web
+from connexion.lifecycle import ConnexionResponse
 
 import wazuh.sca as sca
-from api.encoder import dumps, prettify
+from connexion import request
+from api.controllers.util import json_response
```
OK, the import order was modified in a new commit.

mhamra commented on api/api/controllers/syscollector_controller.py:
```diff
@@ -4,23 +4,23 @@
 
-from aiohttp import web
+from connexion.lifecycle import ConnexionResponse
 
 import wazuh.syscollector as syscollector
-from api.encoder import dumps, prettify
+from connexion import request
+from api.controllers.util import json_response
```
Ok, the import order is fixed in a new commit.

mhamra commented on api/api/controllers/task_controller.py:
```diff
@@ -4,9 +4,10 @@
 
-from aiohttp import web
+from connexion.lifecycle import ConnexionResponse
 
-from api.encoder import dumps, prettify
+from connexion import request
+from api.controllers.util import json_response
```
Ok, the import order is fixed in a new commit.

mhamra commented on api/api/controllers/vulnerability_controller.py:
```diff
@@ -4,9 +4,10 @@
 
-from aiohttp import web
+from connexion.lifecycle import ConnexionResponse
 
-from api.encoder import dumps, prettify
+from connexion import request
+from api.controllers.util import json_response
```
Ok, the import order is fixed in a new commit.

mhamra commented on api/api/error_handler.py:
It is better to maintain the _ and __ notation. If we change the error handlers with "*args, **kwargs" parameters, the functions will not comply with the signature expected by `connexion.AsyncApp.add_error_handler` function.

The parameters _ and __ are conventions indicating that they will not be used in the function, and the linter does raise warnings of unused parameters.

mhamra commented on api/api/middlewares.py:
```diff
@@ -2,50 +2,41 @@
-from connexion.exceptions import OAuthProblem, ProblemException, Unauthorized
-from connexion.problem import problem as connexion_problem
-from secure import SecureHeaders
+import json
+import hashlib
+import time
+import contextlib
+import logging
+
+from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
+from starlette.requests import Request
+from starlette.responses import Response
+from connexion import ConnexionMiddleware
+from connexion.lifecycle import ConnexionRequest
+from secure import Secure, ContentSecurityPolicy, XFrameOptions, Server
```
OK, the import sort order will be fixed in a new commit.

mhamra commented on api/api/controllers/syscheck_controller.py:
```diff
@@ -4,23 +4,23 @@
 
-from aiohttp import web
+from connexion.lifecycle import ConnexionResponse
 
-from api.encoder import dumps, prettify
+from connexion import request
+from api.controllers.util import json_response
```
OK, the imports order will be fixed in a new commit.

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -8,16 +8,46 @@
 import os
 import signal
 import sys
+import asyncio
+import logging
+import logging.config
+import ssl
+
+from connexion import AsyncApp
+from connexion.options import SwaggerUIOptions
+from connexion.exceptions import Unauthorized, HTTPException, BadRequestProblem, ProblemException
+from connexion.middleware import MiddlewarePosition
+import uvicorn
+
+from starlette.middleware.cors import CORSMiddleware
+from content_size_limit_asgi import ContentSizeLimitMiddleware
+
+from jose import JWTError
```
OK, the import order will be fixed in a new commit.

mhamra commented on api/api/controllers/util.py:
```diff
@@ -0,0 +1,29 @@
+XML_CONTENT_TYPE="application/xml; charset=utf-8"
+
+
+def json_response(data: dict, pretty: bool = False) -> ConnexionResponse:
```
The test is found in api.controllers.test.test_controller_util.py. Would you prefer to rename the test_controller_util.py file to test_util.py?

nico-stefani replied:
We decided to rename the file

mhamra commented on api/api/middlewares.py:
```diff
@@ -2,50 +2,41 @@
-    secure_headers.aiohttp(resp)
-    return resp
+# API secure headers
+server = Server().set("Wazuh")
+csp = ContentSecurityPolicy()
+csp.default_src('self')
+xfo = XFrameOptions().deny()
+secure_headers = Secure(server=server, csp=csp, xfo=xfo)
```
I'm not sure; these strings are not used anywhere else. I've adapted the source code found here:

https://github.com/wazuh/wazuh/blob/8cbe309cdd4577853f257183d84ab87b8eff90e6/api/api/middlewares.py#L22-L22

I've made small changes in a new commit.

nico-stefani reviewed (approved):
LGTM! Because we still don't have the embedded Python with the new requirements, it is acceptable that some pipelines fail.

Selutario reviewed (changes requested):
I've tested the behavior of this API, and overall it seems almost the same as before the changes, good job @mhamra!

I found some things that seem wrong though:

### 1. Block login attempts
I am setting this configuration in the `api.yaml`:
```yaml
# Access parameters
access:
 max_login_attempts: 5
 block_time: 20
 max_request_per_minute: 10
```

It does not matter how many times I login using a wrong user/password, it always shows this response:
```JSON
{
  "title": "Unauthorized",
  "detail": "Invalid credentials"
}
```

If then I use the correct credentials, the message changes and warns me about my IP being blocked. However, to keep the same previous behavior, the API should be showing this message as soon as the IP is blocked, it doesn't mind if the credentials are still wrong:
```JSON
{
  "title": "Permission Denied",
  "detail": "Limit of login attempts reached. The current IP has been blocked due to a high number of login attempts",
  "error": 6000
}
```

Also, two logs should be printed everytime I try to login again (one log for the authenticate request and another, for the warning). This is the API log in master branch:
```
2024/01/26 16:54:12 INFO: wazuh 192.168.96.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.322s: 401
2024/01/26 16:54:12 INFO: wazuh 192.168.96.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.155s: 401
2024/01/26 16:54:14 INFO: wazuh 192.168.96.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.152s: 401
2024/01/26 16:54:15 INFO: wazuh 192.168.96.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.155s: 401
2024/01/26 16:54:16 INFO: wazuh 192.168.96.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.150s: 401
2024/01/26 16:54:16 WARNING: IP blocked due to exceeded number of logins attempts: 192.168.96.1
2024/01/26 16:54:16 INFO: wazuh 192.168.96.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.001s: 403
2024/01/26 16:54:17 WARNING: IP blocked due to exceeded number of logins attempts: 192.168.96.1
2024/01/26 16:54:17 INFO: wazuh 192.168.96.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.001s: 403
2024/01/26 16:54:17 WARNING: IP blocked due to exceeded number of logins attempts: 192.168.96.1
2024/01/26 16:54:17 INFO: wazuh 192.168.96.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.001s: 403
2024/01/26 16:54:18 WARNING: IP blocked due to exceeded number of logins attempts: 192.168.96.1
2024/01/26 16:54:18 INFO: wazuh 192.168.96.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.003s: 403
2024/01/26 16:54:27 WARNING: IP blocked due to exceeded number of logins attempts: 192.168.96.1
```

But in this branch, once the IP is blocked, it does not log anything else even if I keep trying to log using wrong credentials:
```
2024-01-26 16:50:10 INFO: wazuh 192.168.80.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.146s: 200
2024-01-26 16:50:11 INFO: wazuh 192.168.80.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.150s: 200
2024-01-26 16:50:12 INFO: wazuh 192.168.80.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.139s: 200
2024-01-26 16:50:13 INFO: wazuh 192.168.80.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.211s: 200
2024-01-26 16:51:08 INFO: wazuh 192.168.80.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.144s: 200
2024-01-26 16:51:19 INFO: wazuh 192.168.80.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.150s: 200
2024-01-26 16:52:07 WARNING: IP blocked due to exceeded number of logins attempts: 192.168.80.1
2024-01-26 16:52:08 WARNING: IP blocked due to exceeded number of logins attempts: 192.168.80.1
2024-01-26 16:52:27 INFO: wazuh 192.168.80.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.142s: 200
2024-01-26 16:52:27 INFO: wazuh 192.168.80.1 "GET /security/user/authenticate" with parameters {} and body {} done in 0.142s: 200
2024-01-26 16:52:28 INFO: wazuh 192.168.80.1 "GET /security/user/authenticate" with parameters {} and body {} done in 0.143s: 200
2024-01-26 16:52:29 INFO: wazuh 192.168.80.1 "GET /security/user/authenticate" with parameters {} and body {} done in 0.140s: 200
```

### 2. Exceed requests per minute
Similar to 1., when max requests per minute are exceeded, it won't allow me to run new requests (which is fine), but it stops logging those requests. All requests should always be logged, in this case, with `status_code` 429 as seen in this log from the API in master branch:
```
2024/01/26 17:05:07 INFO: wazuh 192.168.96.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.567s: 200
2024/01/26 17:05:09 INFO: wazuh 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.106s: 200
2024/01/26 17:05:10 INFO: wazuh 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.027s: 200
2024/01/26 17:05:10 INFO: wazuh 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.026s: 200
2024/01/26 17:05:11 INFO: wazuh 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.029s: 200
2024/01/26 17:05:11 INFO: wazuh 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.029s: 200
2024/01/26 17:05:12 INFO: wazuh 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.028s: 200
2024/01/26 17:05:12 INFO: wazuh 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.034s: 200
2024/01/26 17:05:13 INFO: wazuh 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.027s: 200
2024/01/26 17:05:13 INFO: wazuh 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.027s: 200
2024/01/26 17:05:14 INFO: unknown_user 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.001s: 429
2024/01/26 17:05:14 INFO: unknown_user 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.000s: 429
2024/01/26 17:05:15 INFO: unknown_user 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.000s: 429
2024/01/26 17:05:15 INFO: unknown_user 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.000s: 429
2024/01/26 17:05:16 INFO: unknown_user 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.000s: 429
2024/01/26 17:05:16 INFO: unknown_user 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.002s: 429
2024/01/26 17:05:16 INFO: unknown_user 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.001s: 429
2024/01/26 17:05:17 INFO: unknown_user 192.168.96.1 "GET /agents" with parameters {} and body {} done in 0.001s: 429
```

### 3. Strange error
I ran a request every 0.2 seconds during some time and this error was shown in the API:
- [api.log](https://github.com/wazuh/wazuh/files/14067534/api.log)

I repeated the same in the master branch, and it seems to be working wrong too, but the error is different. Maybe it's worth taking a look to know why this behaves different now:
- [api_master.log](https://github.com/wazuh/wazuh/files/14067553/api_master.log)

mhamra commented on api/api/alogging.py:
It is used by the `set_logging` function. I moved the `set_logging` function to alogging.py

mhamra commented on api/test/integration/test_manager_endpoints.tavern.yaml:
```diff
@@ -1533,9 +1531,9 @@ stages:
         content-type: application/json
     response:
-      status_code: 406
+      status_code: 415
```
Added to spec.yaml

https://github.com/wazuh/wazuh/blob/21f68be574cc75b652dd3af05cbaaac4ab5f6643/api/api/spec/spec.yaml#L644-L652

The error is referenced in these endpoints:
- /lists/files/{filename}: GET, PUT, DELETE
- /rules/files/{filename}:  GET, PUT, DELETE
- /decoders/files/{filename}:  GET, PUT, DELETE

mhamra commented on api/test/integration/test_cdb_list_endpoints.tavern.yaml:
```diff
@@ -732,7 +732,7 @@ stages:
         Authorization: "Bearer {test_login_token}"
     response:
-      status_code: 400
+      status_code: 404
```
Added to spec.yaml

https://github.com/wazuh/wazuh/blob/21f68be574cc75b652dd3af05cbaaac4ab5f6643/api/api/spec/spec.yaml#L644-L652

The error is referenced in these endpoints:
- /lists/files/{filename}: GET, PUT, DELETE
- /rules/files/{filename}:  GET, PUT, DELETE
- /decoders/files/{filename}:  GET, PUT, DELETE

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -7,12 +7,39 @@
+
+from wazuh.rbac.orm import check_database_integrity
+from wazuh.core import pyDaemonModule, common, utils
+from wazuh.core.cluster import __version__, __author__, __wazuh_name__, __licence__
```
I've moved the import statements inside the __main__ block.

https://github.com/wazuh/wazuh/blob/5e6ba1db2bb61a7a1f1ba59224fdcbb919034715/api/scripts/wazuh-apid.py#L297-L360

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -49,80 +78,183 @@ def spawn_authentication_pool():
+    filepath : str
+        File to assign ownership.
+
+    """
```
Fixed

https://github.com/wazuh/wazuh/blob/5e6ba1db2bb61a7a1f1ba59224fdcbb919034715/api/scripts/wazuh-apid.py#L50-L57

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -49,80 +78,183 @@ def spawn_authentication_pool():
 
 
-def start():
+def assign_wazuh_ownership(filepath:str):
```
Fixed

https://github.com/wazuh/wazuh/blob/5e6ba1db2bb61a7a1f1ba59224fdcbb919034715/api/scripts/wazuh-apid.py#L50

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -49,80 +78,183 @@ def spawn_authentication_pool():
+    uvicorn_params : dict
+        uvicorn parameter configuration dictionary.
+
+    """
```
Fixed
https://github.com/wazuh/wazuh/blob/5e6ba1db2bb61a7a1f1ba59224fdcbb919034715/api/scripts/wazuh-apid.py#L66-L73

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -49,80 +78,183 @@ def spawn_authentication_pool():
     except Exception as db_integrity_exc:
-        raise APIError(2012, details=str(db_integrity_exc))
+        raise APIError(2012, details=str(
+            db_integrity_exc)) from db_integrity_exc
```
Fixed

https://github.com/wazuh/wazuh/blob/5e6ba1db2bb61a7a1f1ba59224fdcbb919034715/api/scripts/wazuh-apid.py#L163

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -49,80 +78,183 @@ def spawn_authentication_pool():
+    If another Wazuh API is running, this function will fail because uvicorn server will
+    not be able to create server processes in the same port.
+    The function creates the pool processes, the AsyncApp instance, setups the API spec.yaml, 
+    the middleware classes, the error_handlers, the lifepan, and runs the uvicorn ASGI server.
```
Fixed

https://github.com/wazuh/wazuh/blob/5e6ba1db2bb61a7a1f1ba59224fdcbb919034715/api/scripts/wazuh-apid.py#L153

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -49,80 +78,183 @@ def spawn_authentication_pool():
+    app.add_middleware(WazuhAccessLoggerMiddleware, MiddlewarePosition.BEFORE_VALIDATION)
+    if api_conf['max_upload_size']:
+        app.add_middleware(ContentSizeLimitMiddleware, max_content_size=api_conf['max_upload_size'])
+        app.add_error_handler(ContentSizeExceeded, error_handler.content_size_handler)
```
Fixed

https://github.com/wazuh/wazuh/blob/605ccfe5e23dd33062ecbae2dd46981a1654b562/api/scripts/wazuh-apid.py#L193-L195

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -49,80 +78,183 @@ def spawn_authentication_pool():
+            raise exc from exc
+
+
+def start(params):
```
Fixed

https://github.com/wazuh/wazuh/blob/605ccfe5e23dd33062ecbae2dd46981a1654b562/api/scripts/wazuh-apid.py#L147

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -168,16 +298,177 @@ def exit_handler(signum, frame):
+    returns the logging configuration dictionary that will be used in uvicorn logging
+    configuration.
+    
+    Parameters
```
Fixed

https://github.com/wazuh/wazuh/blob/605ccfe5e23dd33062ecbae2dd46981a1654b562/api/api/alogging.py#L88-L99

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -168,16 +298,177 @@ def exit_handler(signum, frame):
+    Returns
+    -------
+    dict
+        Logging configuraction dictionary.
```
Fixed

https://github.com/wazuh/wazuh/blob/e7665b57c3a2cee25b0d7189062602beb11b2333/api/api/alogging.py#L109-L115

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -168,16 +298,177 @@ def exit_handler(signum, frame):
+            "json" : {
+                '()': 'api.alogging.WazuhJsonFormatter',
+                'style': '%',
+                'datefmt' : "%Y/%m/%d %H:%M:%S"
```
Fixed

https://github.com/wazuh/wazuh/blob/e7665b57c3a2cee25b0d7189062602beb11b2333/api/api/alogging.py#L154

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -168,16 +298,177 @@ def exit_handler(signum, frame):
+        },
+        "loggers": {
+            "wazuh-api" : {"handlers": hdls, "level": log_level, "propagate": False},
+            "start-stop-api" : {"handlers": hdls, "level": 'INFO', "propagate": False}
```
Fixed

https://github.com/wazuh/wazuh/blob/e7665b57c3a2cee25b0d7189062602beb11b2333/api/api/alogging.py#L188-L189

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -168,16 +298,177 @@ def exit_handler(signum, frame):
+                d.update({
+                    'class': 'wazuh.core.wlogging.SizeBasedFileRotatingHandler',
+                    'maxBytes' : max_size,
+                    'backupCount' : 1
```
Fixed

https://github.com/wazuh/wazuh/blob/e7665b57c3a2cee25b0d7189062602beb11b2333/api/api/alogging.py#L200-L201

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -168,16 +298,177 @@ def exit_handler(signum, frame):
+            else:
+                d.update({
+                    'class' : 'wazuh.core.wlogging.TimeBasedFileRotatingHandler',
+                    'when' : 'midnight'
```
Fixed

https://github.com/wazuh/wazuh/blob/e7665b57c3a2cee25b0d7189062602beb11b2333/api/api/alogging.py#L205-L206

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -189,166 +480,41 @@ def exit_handler(signum, frame):
-        print(f"Error when trying to start the Wazuh API. {api_log_error}")
-        sys.exit(1)
+    # Configure uvicorn parameters dictionary
+    uvicorn_params = dict()
```
Fixed

https://github.com/wazuh/wazuh/blob/e7665b57c3a2cee25b0d7189062602beb11b2333/api/scripts/wazuh-apid.py#L364

mhamra commented on api/api/error_handler.py:
```diff
@@ -0,0 +1,241 @@
+
+
+def prevent_bruteforce_attack(request: ConnexionRequest, attempts: int = 5):
+    """Checks that the IPs that are requesting an API token do not do so repeatedly.
```
Fixed

https://github.com/wazuh/wazuh/blob/e7665b57c3a2cee25b0d7189062602beb11b2333/api/api/error_handler.py#L21

mhamra commented on api/api/error_handler.py:
```diff
@@ -0,0 +1,241 @@
+    request : ConnexionRequest
+        Incomming request.
+    exc: Unauthorized
+        Raised exception.
```
Fixed

https://github.com/wazuh/wazuh/blob/e7665b57c3a2cee25b0d7189062602beb11b2333/api/api/error_handler.py#L69-L70

mhamra commented on api/api/error_handler.py:
```diff
@@ -0,0 +1,241 @@
+        Incomming request.
+        Unnamed parameter not used.
+    exc: BadRequestProblem
+        Raised exception.
```
Fixed

https://github.com/wazuh/wazuh/blob/e7665b57c3a2cee25b0d7189062602beb11b2333/api/api/error_handler.py#L104-L108

mhamra commented on api/api/error_handler.py:
```diff
@@ -0,0 +1,241 @@
+    _ : ConnexionRequest
+        Incomming request.
+        Unnamed parameter not used.
+    __: JWTError
```
Fixed

https://github.com/wazuh/wazuh/blob/e7665b57c3a2cee25b0d7189062602beb11b2333/api/api/error_handler.py#L161

mhamra commented on api/api/error_handler.py:
```diff
@@ -0,0 +1,241 @@
+        Incomming request.
+        Unnamed parameter not used.
+    exc: ContentSizeExceeded
+        Raised exception.
```
Fixed

https://github.com/wazuh/wazuh/blob/e7665b57c3a2cee25b0d7189062602beb11b2333/api/api/error_handler.py#L223-L227

mhamra commented on api/api/middlewares.py:
```diff
@@ -153,83 +111,119 @@ async def check_rate_limit(
+    msg = f'Listening on {configuration.api_conf["host"]}:{configuration.api_conf["port"]}.'
+    start_stop_logger.info(msg)
+    yield
+    start_stop_logger.info('Shutdown wazuh-apid server.')
```
Fixed

https://github.com/wazuh/wazuh/blob/e7665b57c3a2cee25b0d7189062602beb11b2333/api/api/middlewares.py#L227-L229

mhamra commented on api/api/middlewares.py:
```diff
@@ -153,83 +111,119 @@ async def check_rate_limit(
-            max_request_per_minute
-        )
+    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
+        """"Checks request limits per minute"""
```
Fixed

https://github.com/wazuh/wazuh/blob/e43d092cab72c08c4fe55a6679ffc46a16c901b2/api/api/middlewares.py#L123

mhamra commented on api/api/middlewares.py:
```diff
@@ -153,83 +111,119 @@ async def check_rate_limit(
+    """Middleware to log custom Access messages."""
+
+    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
+        """Logs Wazuh access information.
```
Fixed

https://github.com/wazuh/wazuh/blob/99d97a092434d60cf798a4666b11e6ab9a497a0d/api/api/middlewares.py#L151

Selutario commented on api/scripts/wazuh-apid.py:
```diff
@@ -169,19 +275,47 @@ def exit_handler(signum, frame):
     pyDaemonModule.delete_pid(API_MAIN_PROCESS, api_pid)
 
 
+def add_debug2_log_level_and_error():
+    """Add a new debug level used by wazuh api and framework."""
+
+    logging.DEBUG2 = 6
+
+    def debug2(self, message, *args, **kws):
+        if self.isEnabledFor(logging.DEBUG2):
+            self._log(logging.DEBUG2, message, args, **kws)
+
+    def error(self, msg, *args, **kws):
+        if self.isEnabledFor(logging.ERROR):
+            if 'exc_info' not in kws:
+                kws['exc_info'] = self.isEnabledFor(logging.DEBUG2)
+            self._log(logging.ERROR, msg, args, **kws)
+
+    logging.addLevelName(logging.DEBUG2, "DEBUG2")
+
+    logging.Logger.debug2 = debug2
+    logging.Logger.error = error
```
There seems to be something wrong. When using this api configuration:
```YAML
 logs:
  level: "debug2"
  format: "plain,json"
  max_size:
    enabled: True
    size: "1M"
```

This error is shown:
```
# /var/ossec/bin/wazuh-apid -f
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.10/logging/config.py", line 619, in configure
    self.configure_logger(name, loggers[name])
  File "/var/ossec/framework/python/lib/python3.10/logging/config.py", line 796, in configure_logger
    self.common_logger_config(logger, config, incremental)
  File "/var/ossec/framework/python/lib/python3.10/logging/config.py", line 781, in common_logger_config
    logger.setLevel(logging._checkLevel(level))
  File "/var/ossec/framework/python/lib/python3.10/logging/__init__.py", line 198, in _checkLevel
    raise ValueError("Unknown level: %r" % level)
ValueError: Unknown level: 'DEBUG2'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/ossec/api/scripts/wazuh-apid.py", line 388, in <module>
    logging.config.dictConfig(uvicorn_params['log_config'])
  File "/var/ossec/framework/python/lib/python3.10/logging/config.py", line 811, in dictConfig
    dictConfigClass(config).configure()
  File "/var/ossec/framework/python/lib/python3.10/logging/config.py", line 621, in configure
    raise ValueError('Unable to configure logger '
ValueError: Unable to configure logger 'wazuh-api'
```

mhamra replied:
Fixed. The debug level must be created before configuring the logs.

https://github.com/wazuh/wazuh/blob/3672b5ef17f00770e66fa9433eca37834749e870/api/scripts/wazuh-apid.py#L388-L391

Selutario reviewed (changes requested):
This looks better now. I found some additional problems though:

### Too many requests
The response when `429 - Too many requests` is different than in master, it does not contain the `remediation` field.
#### master
```JSON
{
  "title": "Too Many Requests",
  "detail": "Maximum number of requests per minute reached",
  "remediation": "This limit can be changed in api.yaml file. More information here: https://documentation.wazuh.com/4.9/user-manual/api/configuration.html#configuration-file",
  "error": 6001
}
```

#### upgrade/16504-upgrade-connexion-to-3.0
```JSON
{
  "title": "Maximum number of requests per minute reached",
  "error": 6001
}
```

### Permission denied
When a user with readonly role (for example) tries to something forbidden like adding an agent, the response is correct:
```JSON
{
  "title": "Permission Denied",
  "detail": "Permission denied: Resource type: *:*",
  "remediation": "Please, make sure you have permissions to execute the current request. For more information on how to set up permissions, please visit https://documentation.wazuh.com/4.9/user-manual/api/rbac/configuration.html",
  "dapi_errors": {
    "master-node": {
      "error": "Permission denied: Resource type: *:*"
    }
  },
  "error": 4000
}
```

However, in the master branch, the request is logged in the `api.log`:
```
2024/02/13 12:24:41 INFO: test 172.20.0.1 "POST /agents" with parameters {} and body {"name": "test_agent9", "ip": "any"} done in 0.010s: 403
```
In `upgrade/16504-upgrade-connexion-to-3.0` nothing is printed when permission was denied.

mhamra commented:
## UPDATE
1. Too many requests
Executing the `test_DOS_blocking_system.py`, this is the response shown while debugging the test. The results obtained for the status code 429 are visible in the image.

![image](https://github.com/wazuh/wazuh/assets/2949519/aeb71cd9-4162-4130-9cb7-182fb9c9c5ab)

2. Permission denied
This is the log of the `test_rbac_black_agent_endpoints.tavern.yaml` AIT. Permission error 403 is now logged by the `wazuh-apid.py` script

```
root@wazuh-master:/# tail -f /var/ossec/logs/api.log 
2024/02/15 14:37:58 INFO: HTTPS is enabled but cannot find the private key and/or certificate. Attempting to generate them
2024/02/15 14:37:58 INFO: Generated private key file in WAZUH_PATH/api/configuration/ssl/server.key
2024/02/15 14:37:58 INFO: Generated certificate file in WAZUH_PATH/api/configuration/ssl/server.crt
2024/02/15 14:37:58 INFO: Checking RBAC database integrity...
2024/02/15 14:37:58 INFO: RBAC database not found. Initializing
2024/02/15 14:38:01 INFO: /var/ossec/api/configuration/security/rbac.db database created successfully
2024/02/15 14:38:01 INFO: RBAC database integrity check finished successfully
2024/02/15 14:38:05 INFO: Listening on 0.0.0.0:55000.
2024/02/15 14:38:05 INFO: testing 127.0.0.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.016s: 401
2024/02/15 14:38:07 INFO: testing 127.0.0.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.298s: 200
2024/02/15 14:40:55 INFO: testing 172.22.0.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.180s: 200
2024/02/15 14:40:55 INFO: testing 172.22.0.1 "GET /agents" with parameters {} and body {} done in 0.047s: 200
2024/02/15 14:40:55 INFO: testing 172.22.0.1 "GET /agents" with parameters {"agents_list": "002,004"} and body {} done in 0.023s: 403
2024/02/15 14:40:56 INFO: testing 172.22.0.1 "GET /agents" with parameters {"agents_list": "001,003,004,005,006"} and body {} done in 0.024s: 200
2024/02/15 14:40:56 INFO: testing 172.22.0.1 "GET /agents" with parameters {"agents_list": "001,003,007,011"} and body {} done in 0.024s: 200
```

mhamra commented on api/api/authentication.py:
```diff
@@ -28,6 +28,7 @@
 from wazuh.rbac.orm import AuthenticationManager, TokenManager, UserRolesManager
 from wazuh.rbac.preprocessor import optimize_resources
 
+INVALID_TOKEN = "Invalid token"
```
Modified:
https://github.com/wazuh/wazuh/blob/7d4df6ede3e7c958b79652773b884f27b97ccec5/api/api/authentication.py#L31-L32

mhamra commented on api/api/authentication.py:
```diff
@@ -292,8 +292,8 @@ def decode_token(token: str) -> dict:
         if payload['rbac_policies']['rbac_mode'] != current_rbac_mode \
                 or (payload['exp'] - payload['nbf']) != current_expiration_time:
-            raise Unauthorized
+            raise Unauthorized("Token Expired")
```
Modified:
https://github.com/wazuh/wazuh/blob/7d4df6ede3e7c958b79652773b884f27b97ccec5/api/api/authentication.py#L296

mhamra commented on api/api/controllers/agent_controller.py:
```diff
@@ -6,34 +6,34 @@
+from api.util import deprecate_endpoint, parse_api_param, raise_if_exc, remove_nones_to_dict
 from api.validator import check_component_configuration_pair
-from wazuh import agent, stats
+from wazuh import agent
```
Modified:
https://github.com/wazuh/wazuh/blob/7d4df6ede3e7c958b79652773b884f27b97ccec5/api/api/controllers/agent_controller.py#L19

mhamra commented on api/api/controllers/agent_controller.py:
```diff
@@ -6,34 +6,34 @@
 from wazuh.core.cluster.dapi.dapi import DistributedAPI
 from wazuh.core.common import DATABASE_LIMIT
 from wazuh.core.results import AffectedItemsWazuhResult
+from wazuh import stats
```
Extra import removed.

mhamra commented on api/api/middlewares.py:
```diff
@@ -55,94 +55,114 @@ async def set_secure_headers(request, handler):
+    if 'key' in body and '/agents' in path:
+        body['key'] = '****'
+
+    # Get the user name from the request. If it is not found in the context, 
```
Fixed:
https://github.com/wazuh/wazuh/blob/7d4df6ede3e7c958b79652773b884f27b97ccec5/api/api/middlewares.py#L80

mhamra commented on api/api/alogging.py:
```diff
@@ -201,3 +83,192 @@ def add_fields(self, log_record: collections.OrderedDict, record: logging.LogRec
+                'filename': f"{log_filepath}.log",
+                'formatter': 'log',
+                'filters': ['plain-filter'],
+            }
```
Fixed. Now the log files can be generated even when the `wazuh-apid.py` script is launched in foreground mode.

https://github.com/wazuh/wazuh/blob/7d4df6ede3e7c958b79652773b884f27b97ccec5/api/api/alogging.py#L115-L129

mhamra commented on api/api/alogging.py:
```diff
@@ -201,3 +83,192 @@ def add_fields(self, log_record: collections.OrderedDict, record: logging.LogRec
+            "log": {
+                "()": "uvicorn.logging.DefaultFormatter",
+                "fmt": "%(asctime)s %(levelname)s: %(message)s",
+                "datefmt": "%Y-%m-%d %H:%M:%S",
```
Fixed. Note: The `wazuh-apid.py` script does not log the Uvicorn server access log messages. It is possible to enable the logging, but it is not useful.

https://github.com/wazuh/wazuh/blob/7d4df6ede3e7c958b79652773b884f27b97ccec5/api/api/alogging.py#L151

mhamra commented on api/api/alogging.py:
```diff
@@ -201,3 +83,192 @@ def add_fields(self, log_record: collections.OrderedDict, record: logging.LogRec
+        json_info['body'] = body
+
+    log_info += f'with parameters {json.dumps(query)} and body'\
+            f' {json.dumps(body)} done in {elapsed_time:.3f}s: {status}'
```
Fixed.

https://github.com/wazuh/wazuh/blob/7d4df6ede3e7c958b79652773b884f27b97ccec5/api/api/alogging.py#L266-L267

mhamra commented on api/api/middlewares.py:
```diff
@@ -55,94 +55,114 @@ async def set_secure_headers(request, handler):
+
+    custom_logging(user, host, method,
+                    path, query, body, time_diff, response.status_code,
+                    hash_auth_context=hash_auth_context, headers=headers)
```
Fixed.
https://github.com/wazuh/wazuh/blob/7d4df6ede3e7c958b79652773b884f27b97ccec5/api/api/middlewares.py#L105-L106

mhamra commented on api/api/middlewares.py:
```diff
@@ -55,94 +55,114 @@ async def set_secure_headers(request, handler):
     """
     global ip_block, ip_stats
+    access_conf = configuration.api_conf['access']
+    block_time=access_conf['block_time']
```
Fixed:
https://github.com/wazuh/wazuh/blob/7d4df6ede3e7c958b79652773b884f27b97ccec5/api/api/middlewares.py#L126

mhamra commented on api/api/middlewares.py:
```diff
@@ -153,83 +173,102 @@ async def check_rate_limit(
+        """"Update and check if the client IP is locked."""
+
+        check_blocked_ip(request)
+        return await call_next(request)
```
Fixed. The middleware now checks if the request is from authentication endpoints.

https://github.com/wazuh/wazuh/blob/7d4df6ede3e7c958b79652773b884f27b97ccec5/api/api/middlewares.py#L210-L215

Selutario replied:
There seems to be a problem somewhere related, I added more details here:
- https://github.com/wazuh/wazuh/pull/20428#discussion_r1494456753

mhamra commented on api/scripts/wazuh-apid.py:
```diff
@@ -190,166 +324,86 @@ def exit_handler(signum, frame):
-        print(f"Error when trying to start the Wazuh API. {api_log_error}")
+        uvicorn_params['log_config'] = set_logging(log_filepath=API_LOG_PATH,
+                                                log_level=api_conf['logs']['level'].upper(),
+                                                foreground_mode=args.foreground)
```
Fixed.
https://github.com/wazuh/wazuh/blob/7d4df6ede3e7c958b79652773b884f27b97ccec5/api/scripts/wazuh-apid.py#L375-L377

mhamra commented on api/api/spec/spec.yaml:
```diff
@@ -12036,6 +12062,8 @@ paths:
           $ref: '#/components/responses/WrongContentTypeResponse'
         '413':
           $ref: '#/components/responses/RequestTooLargeResponse'
+        '415':
+          $ref: '#/components/responses/UnsupportedMediaTypeResponse'
```
I would suggest to validate this in another issue. 

The `Body.validate_content_type` class method raises a `WazuhNotAcceptable(6002)`. This exception is now handled by the generic error handler `problem_error_handler`. I think that it will be better to create a custom exception inherited from `ProblemException` to generate the same response and notify the content_type error.

mhamra replied:
This request generates a 415 error response. The header `Content-Type` does not meet the defined type in the `spec.yaml`, which should be `application/octet-stream`. 

```console
curl -k -X PUT "https://localhost:55000/manager/configuration?wait_for_complete=true&pretty=true" -H "Authorization: Bearer $TOKEN" -H "Content-Type: application/json"
{
   "title": "Unsupported Media Type",
   "detail": "Invalid Content-type (application/json), expected ['application/octet-stream']"
}
```
Every endpoint with the `content-type` defined in the spec will generate a 415 error if the request defines another `content-type`.

Selutario replied:
I tried again and it returns 415 as you said. Maybe there was something wrong with my mappings.
![image](https://github.com/wazuh/wazuh/assets/23361101/31b8d1da-a57d-42f8-98cd-f357f7825f82)

Selutario commented on api/api/spec/spec.yaml:
```diff
@@ -11532,6 +11552,8 @@ paths:
         '403':
           $ref: '#/components/responses/PermissionDeniedResponse'
+        '404':
+          $ref: '#/components/responses/InvalidUriResponse'          
```
Mmm I see. The `%2F` code is a URL encoded `/`, so, as it was shown in the logs, it would be equivalent to doing:
- `GET /lists/files/te/st`

Endpoints never expect `/` in their path parameters, as they would be considered different (unexistent) endpoints. This same thing would happen with any endpoint actually, not only with those expecting a path parameter:
```JSON
# curl -k -X GET "https://localhost:55000/agents/test" -H "Content-Type:application/json" -H "Authorization: Bearer $TOKEN"
{"title": "Not Found", "detail": "404: Not Found"}
``` 

If instead of including an encoded `/` we include a `-` (`%2D`), it works correctly:
```JSON
# curl -k -X GET "https://localhost:55000/lists/files/te%2Dst?pretty=true" -H "Content-Type:application/json" -H "Authorization: Bearer $TOKEN"
{
   "data": {
      "affected_items": [],
      "total_affected_items": 0,
      "total_failed_items": 1,
      "failed_items": [
         {
            "error": {
               "code": 1802,
               "message": "Lists file not found",
               "remediation": "Please, use `GET /lists/files` to find all available lists"
            },
            "id": [
               "te-st"
            ]
         }
      ]
   },
   "message": "No list was returned",
   "error": 1
}
```

The conclusion is that the behavior of these endpoints is not different from the rest, so we do not need to specify 404 among their possible responses.

mhamra replied:
Understood; I removed error 404 from the endpoints in the `specs.yaml`.

Selutario commented on api/api/middlewares.py:
```diff
@@ -153,83 +174,99 @@ async def check_rate_limit(
+            return await call_next(request)
+
+
+class CheckBlockedIP(BaseHTTPMiddleware):
+    """Rate Limits Middleware."""
+
+    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
+        """"Update and check if the client IP is locked."""
+        if request.url.path in {LOGIN_ENDPOINT, RUN_AS_LOGIN_ENDPOINT} \
+           and request.method in {'GET', 'POST'}:
+            check_blocked_ip(request)
+        return await call_next(request)
```
With the configuration below:
```YAML
# Access parameters
access:
 max_login_attempts: 5
 block_time: 20
 max_request_per_minute: 10
```

I'm getting some errors when the IP should be blocked after failing 5 auth attempts:
```
2024/02/19 12:10:33 INFO: Starting API in foreground
2024/02/19 12:10:33 INFO: Checking RBAC database integrity...
2024/02/19 12:10:33 INFO: /var/ossec/api/configuration/security/rbac.db file was detected
2024/02/19 12:10:33 INFO: RBAC database integrity check finished successfully
2024/02/19 12:10:38 INFO: Listening on 0.0.0.0:55000.
2024/02/19 12:10:41 INFO: wazuh 192.168.48.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.221s: 401
2024/02/19 12:10:41 INFO: wazuh 192.168.48.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.168s: 401
2024/02/19 12:10:42 INFO: wazuh 192.168.48.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.164s: 401
2024/02/19 12:10:43 INFO: wazuh 192.168.48.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.166s: 401
2024/02/19 12:10:44 INFO: wazuh 192.168.48.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.172s: 401
2024/02/19 12:10:44 ERROR: Exception in ASGI application
  + Exception Group Traceback (most recent call last):
  |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/_utils.py", line 82, in collapse_excgroups
  |     yield
  |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 190, in __call__
  |     async with anyio.create_task_group() as task_group:
  |   File "/var/ossec/framework/python/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 678, in __aexit__
  |     raise BaseExceptionGroup(
  | exceptiongroup.ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/errors.py", line 164, in __call__
    |     await self.app(scope, receive, _send)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 189, in __call__
    |     with collapse_excgroups():
    |   File "/var/ossec/framework/python/lib/python3.10/contextlib.py", line 153, in __exit__
    |     self.gen.throw(typ, value, traceback)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/_utils.py", line 88, in collapse_excgroups
    |     raise exc
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 191, in __call__
    |     response = await self.dispatch_func(request, call_next)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/api/middlewares.py", line 237, in dispatch
    |     response = await call_next(request)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 165, in call_next
    |     raise app_exc
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 151, in coro
    |     await self.app(scope, receive_or_disconnect, send_no_error)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/connexion/middleware/exceptions.py", line 101, in __call__
    |     await super().__call__(scope, receive, send)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
    |     await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/_exception_handler.py", line 63, in wrapped_app
    |     response = await handler(conn, exc)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/connexion/middleware/exceptions.py", line 37, in wrapper
    |     response = await handler(request, exc)  # type: ignore
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/api/error_handler.py", line 196, in problem_error_handler
    |     problem.update(exc.ext if exc.ext else {})
    | TypeError: 'ConnexionRequest' object is not subscriptable
    +------------------------------------

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 189, in __call__
    with collapse_excgroups():
  File "/var/ossec/framework/python/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/_utils.py", line 88, in collapse_excgroups
    raise exc
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 191, in __call__
    response = await self.dispatch_func(request, call_next)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/api/middlewares.py", line 237, in dispatch
    response = await call_next(request)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 165, in call_next
    raise app_exc
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 151, in coro
    await self.app(scope, receive_or_disconnect, send_no_error)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/connexion/middleware/exceptions.py", line 101, in __call__
    await super().__call__(scope, receive, send)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/_exception_handler.py", line 63, in wrapped_app
    response = await handler(conn, exc)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/connexion/middleware/exceptions.py", line 37, in wrapper
    response = await handler(request, exc)  # type: ignore
  File "/var/ossec/framework/python/lib/python3.10/site-packages/api/error_handler.py", line 196, in problem_error_handler
    problem.update(exc.ext if exc.ext else {})
TypeError: 'ConnexionRequest' object is not subscriptable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.10/site-packages/uvicorn/protocols/http/h11_impl.py", line 408, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/var/ossec/framework/python/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/connexion/apps/abstract.py", line 284, in __call__
    return await self.middleware(scope, receive, send)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/connexion/middleware/main.py", line 501, in __call__
    await self.app(scope, receive, send)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/errors.py", line 181, in __call__
    await response(scope, receive, send)
TypeError: 'coroutine' object is not callable
```

mhamra replied:
I fixed the problem in a new commit. I reproduced the failure by running the integration `test_api/test_config/test_bruteforce_blocking_system`. The test now pass OK.

```
root@b843e90c07fb:/wazuh/tests/integration/test_api# pytest -vv test_config/test_bruteforce_blocking_system/
============================================================================================= test session starts ==============================================================================================
platform linux -- Python 3.10.12, pytest-7.1.2, pluggy-1.4.0 -- /usr/bin/python3
cachedir: .pytest_cache
metadata: {'Python': '3.10.12', 'Platform': 'Linux-5.15.133.1-microsoft-standard-WSL2-x86_64-with-glibc2.35', 'Packages': {'pytest': '7.1.2', 'pluggy': '1.4.0'}, 'Plugins': {'html': '3.1.1', 'metadata': '3.1.1'}}
rootdir: /wazuh/tests/integration, configfile: pytest.ini
plugins: html-3.1.1, metadata-3.1.1
collected 1 item

test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py::test_bruteforce_blocking_system[BRUTE_FORCE_BLOCKING_TIME] PASSED                                                        [100%]

============================================================================================== 1 passed in 28.20s ==============================================================================================
```

mhamra commented:
### UPDATE

- I rebased the PR to master solving the conflicts and pushing.
- I've shown how to generate error 415 in the `PUT /manager/configuration` endpoint.
- I've implemented the `pretty` parameter in the error handlers.
- I've committed the changes to the repository.

Selutario commented on api/api/spec/spec.yaml:
```diff
@@ -7605,8 +7615,6 @@ paths:
           $ref: '#/components/responses/UnauthorizedResponse'
         '403':
           $ref: '#/components/responses/PermissionDeniedResponse'
-        '404':
-          $ref: '#/components/responses/ResourceNotFoundResponse'
```
I would keep previous references to `#/components/responses/ResourceNotFoundResponse`, since there are some endpoints where 404 is actually returned:
```JSON
# DELETE /agents/group?agents_list=000&group_id=test
{
  "title": "Resource Not Found",
  "detail": "The group does not exist",
  "type": "about:blank",
  "remediation": "Please, use `GET /agents/groups` to find all available groups",
  "dapi_errors": {
    "master-node": {
      "error": "The group does not exist"
    }
  },
  "error": 1710
}
``` 

What I meant [here](https://github.com/wazuh/wazuh/pull/20428#discussion_r1494407416) was to remove only the new 404 codes introduced in this PR, or in other words, only `InvalidUriResponse`.

mhamra replied:
I Reverted all the removed lines found in the `4.9.`0 branch like the following:
```
        '404':
          $ref: '#/components/responses/ResourceNotFoundResponse'
```

mhamra replied:
I removed the lines from the `spec.yaml` file with the `'404'` response code that I've added unnecessarily.

Selutario commented on api/api/spec/spec.yaml:
```diff
@@ -7648,8 +7656,6 @@ paths:
           $ref: '#/components/responses/UnauthorizedResponse'
         '403':
           $ref: '#/components/responses/PermissionDeniedResponse'
-        '404':
-          $ref: '#/components/responses/ResourceNotFoundResponse'
```
The same as above, for all the previous existing 404 ResourceNotFoundResponse codes.

mhamra replied:
I Reverted all the removed lines found in the `4.9.`0 branch like the following:
```
        '404':
          $ref: '#/components/responses/ResourceNotFoundResponse'
```

Selutario reviewed (changes requested):
Almost ready, just one last thing

Selutario reviewed (approved):
LGTM. Some of the failing checks are unrelated to these changes, while others are failing because multiple dependencies were changed and need to be updated in:
- https://github.com/wazuh/wazuh/issues/20795

Since this PR is part of an epic issue, all this will still be reviewed in the future.
